### ISSUE TITLE
bitcoind can't find the datadir on windows

### ISSUE THREAD

**Issue by ptschip** [gandrewstone/bitcoin#332]
This is on the current dev branch...when we launch bitcoind and we have previously bitcoin it should find our datadir if it's not in the default location.  There should be a reg setting that it's referencing from previous installations however how it doesn't work anymore since the latest merges to dev.  Not sure which merge was the problem.

**Comment by ptschip** [comment]
Now i have to explicitly tell it where the datadir is , if it's not the default location.

**Comment by Justaphf** [comment]
This may be broader than just the datadir, at some point after building dev I lost my preference that hitting X just hides the client instead of shutting down.  I had to re-specify that preference, though once I did it was fixed, even if I switch back and forth between dev and release builds. Makes me think the issue may be a change in on of the registry key roots.

**Comment by Justaphf** [comment]
I think I see what may be the issue, I ended up with 3 different registry keys, not sure what commits caused the changes.  I haven't run an installer since v0.12.0 so additional changes came from running the client only.

In the image below, the Bitcoin\Bitcoin-Qt folder has a lot more keys, including the data directory.  The BitcoinUnlimited\Bitcoin-Qt folder has a sub-set of the keys, and is missing the data dir (key name `strDataDir`).

![image](https://cloud.githubusercontent.com/assets/6402604/23736382/e4f9f7e2-0459-11e7-8d83-23cf5fc4394e.png)

**Comment by martinrlie** [comment]
Upgraded from 1.0.0.0 to 1.0.1.0 (on Win10), thinking it was a drop-in replacement, but then it started syncing from genesis.

Apparently, 1.0.0.0 uses the the HKCU\SOFTWARE\Bitcoin\Bitcoin-Qt\strDataDir value, while 1.0.1.0 ignores it.

(I also have the additional "HKCU\SOFTWARE\BitcoinUnlimited" and "HKCU\SOFTWARE\Bitcoin Unlimited (64-bit)" keys that you can see in @Justaphf's screenshot. Time to pick one and stay with it, including facilitating a more graceful switch-over from "old" settings? ;))

**Comment by ptschip** [comment]
@justaphf  any chance you have time to come up with a good solution for
all this?...not sure how this problem got introduced but we obviously
need a fix...

**Comment by ptschip** [comment]
@justaphf  now that some change has been made, perhaps best just to
leave it ...it probably only effects users that have their blockchain in
a non-standard location?  I think that's what i was finding anyway.  I
have no trouble with my nodes except the one where i put the blockchain
on different drive.    I would think doing a fresh install using the
installer would fix that?

**Comment by Justaphf** [comment]
@ptschip I am currently swamped at work so I don't think I'll have time to look this week.  There were some new devs in the slack that seemed to be looking for some small tasks to work on so maybe they can pick this up?

For this one, probably also want to look and see if it is further affected by #361, as I have a feeling this issue may have been caused by changes to some of the product strings, though I'm not sure why it just showed up now, not back in 12.1 when the majority of the strings were modified.

**Comment by nimchak** [comment]
@Justaphf  nice catch.  After I created the strDataDir in Computer\HKEY_CURRENT_USER\Software\BitcoinUnlimited\Bitcoin-Qt it all worked as expected.

**Comment by Justaphf** [comment]
@ptschip Was able to trace this down to 75c7b7fa (EDIT: also affects release branch, 7395d35a)
On Windows, `QSettings` writes to the registry under HKEY_CURRENT_USER\SOFTWARE\ `QAPP_ORG_NAME` \ `QAPP_APP_NAME_DEFAULT`, and since the value of `QAPP_ORG_NAME` changed we get the different registry key.

I don't think the correct solution is to revert this value, so I think what would be best is to introduce some settings migration logic that can copy the settings from a different org base to the new org base.  Would need to detect if we've already performed the migration so as to not keep overwriting the settings, should be as simple as adding a new `fHasMigrated` setting which we check for first, before attempting to migrate old settings.

I was able to do a quick test which worked as I expected on Windows as a proof-of-concept.  This is very rough and would need clean-up and testing on other platforms to ensure it works everywhere.  This only affects the Qt client as everything is controlled via `QSettings`.  Since I didn't have a lot of time to look into it I put any question areas to look into as inline comments.

In src/qt/bitcoin.cpp added the following test function
```
//REVISIT: We should pass in both the source and destination org names
void MigrateSettings(const char* orgName)
{
    //Reference the new registry location first
    QSettings sink(orgName, QAPP_APP_NAME_DEFAULT);

    //TODO: We should check to see if we actually need to migrate and skip out if we don't

    //Reference the old registry location (REVISIT: May create the root if it doesn't already exist)
    QSettings source("Bitcoin", QAPP_APP_NAME_DEFAULT);

    //Get the list of all keys in the source location (REVISIT: Not sure how this handles sub-keys)
    QStringList keys = source.allKeys();

    //Loop through all keys in the source registry location
    //NOTE: Loop may not handle sub-keys correctly w/o modification (though currently there aren't any sub-keys anyway)
    Q_FOREACH(const QString& key, keys)
    {
        //copy every setting in source to sink, even if the key is empty
        //NOTE: this will overwrite any existing values in the destination that match the source key name
        sink.setValue(key, source.value(key));
    }

    //REVISIT: Not sure how Qt handles access issues with registry, so this may need some try-catch logic wrapped around it.
}

```
Test call: `MigrateSettings("BitcoinTest");//proof-of-concept` from inside of `main()`, just before setting the Org Name of the application, which correctly created an exact copy of Bitcoin\Bitcoin-Qt to BitcoinTest\Bitcoin-Qt.

NOTE: My time is very limited for the next 3 weeks, so if someone wants to take this research and run with it to create a finished PR, please do.

**Comment by sickpig** [comment]
Closed by #443
